spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -348,17 +425,71 @@ function onRecvPacket(packet: Packet) {
     } else {
       // since we are receiving to a new sink zone we prepend the prefix to the trace
       prefix = "{packet.destPort}/{packet.destChannel}/"
-      newTrace = append([]string{prefix}, token.trace...)
-      onChainDenom = constructOnChainDenom(newTrace, token.denom)
+      onChainTrace = append([]string{prefix}, token.trace...)
+      onChainDenom = constructOnChainDenom(onChainTrace, token.denom)
       // sender was source, mint vouchers to receiver (assumed to fail if balance insufficient)
-      err = bank.MintCoins(data.receiver, onChainDenom, token.amount)
+      err = bank.MintCoins(receiver, onChainDenom, token.amount)
       if (err !== nil) {
         ack = FungibleTokenPacketAcknowledgement{false, "mint coins failed"}
         // break out of for loop on first error
         break
       }
     }
+
+    // add the received token to the received tokens list
+    recvToken = Token{
+      denom: token.denom,
+      trace: onChainTrace,
+      amount: token.amount,
+    }
+    receivedTokens = append(receivedTokens, recvToken)
   }
+
+  // if there is an error ack return immediately and do not forward further
+  if !ack.Success() {
+    return ack
+  }
+
+  // if acknowledgement is successful and forwarding path set
+  // then start forwarding
+  if len(forwardingPath.hops) > 0 {
+    //check that next channel supports token forwarding
+    channel = provableStore.get(channelPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID))
+    if channel.version != "ics20-2" && len(forwardingPath.hops) > 1 {
+      ack = FungibleTokenPacketAcknowledgement(false, "next hop in path cannot support forwarding onward")
+      return ack
+    }
+    memo = ""
+    nextForwardingPath = ForwardingPath{
+      hops: forwardingPath.hops[1:]
+      memo: forwardingPath.memo
+    }
+    if forwardingPath.hops == 1 {
+      // we're on the last hop, we can set memo and clear
+      // the next forwardingPath
+      memo = forwardingPath.memo
+      nextForwardingPath = nil
+    }
+    // send the tokens we received above to the next port and channel
+    // on the forwarding path
+    // and reduce the forwardingPath by the first element
+    nextPacketSequence = sendFungibleTokens(
+      receivedTokens,
+      receiver, // sender of next packet
+      finalReceiver, // receiver of next packet
+      memo,
+      nextForwardingPath,
+      forwardingPath.hops[0].portID,
+      forwardingPath.hops[0].channelID,
+      Height{},
+      currentTime() + DefaultHopTimeoutPeriod,
+    )
+    // store packet for future sending ack
+    privateStore.set(packetForwardPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID, nextPacketSequence), packet)
```

Comment:
Note that we have an incompatibility in the number of parameters used with the `packetForwardPath` in the : 
- `OnRecvPacket`  :  we use `packetForwardPath(forwardingPath.hops[0].portID, forwardingPath.hops[0].channelID, nextPacketSequence)`. Here we include nextPacketSequence as a key for the path. 

- `onAcknowledgePacket`&& `onTimeoutPacket` :  we use `packetForwardPath(packet.sourcePort, packet.sourceChannel)` without using the nextPacketSequence as a discriminator for the key. 

The result is that the previous packet in `onAcknowledgePacket`&& `onTimeoutPacket` is always nil. 

Since, unless we store it somewhere, should be not possible to retrieve the nextPacketSequence in `onAcknowledgePacket` && `onTimeoutPacket`, a couple of possible solutions could be: 
1. remove nextPacketSequence in the key for packetForwardPath, 
2. remove nextPacketSequence in the key for packetForwardPath, but save it together with packet in store. 

Do you see a possible better approach?